File: 02_Dawn/pico/vector3d.c
#include "vector3d.h"
#include "mathtab.h"

int v3_angle_x = 0;
int v3_angle_y = 0;
int v3_angle_z = 0;
int v3_translation = 3000;
int v3_zoom = 1000;

/* The assembly's `lsr #5` converts a 16-bit angle to a 0..2047 index into a
 * sine table of length 2048 (the original keeps multiple copies to avoid a
 * mask). We use a single 1024 table and mask — equivalent up to a /2 in
 * granularity, indistinguishable visually. */
static inline int angle_to_sin_idx(int angle)
{
    return (angle >> 6) & SIN_TAB_MASK;   /* 16-bit → 1024 */
}

/* sin_tab values peak at ~32255 (~31.5 K). Treating that as 1.0 in 16.16
 * means a multiply by sin then >> 15 gives a signed-fixed result roughly
 * equivalent to sin * x with sin in [-1, 1]. We use shifts that match the
 * assembly's fixed-point flow (no calibration constants needed). */
static inline int fixmul_sin(int v, int s)
{
    return (v * s) >> 15;
}

void v3_rotate_project(const vec3_t *in, screen_pt_t *out)
{
    const int x0 = in->x;
    const int y0 = in->y;
    const int z0 = in->z;

    const int ay = angle_to_sin_idx(v3_angle_y);
    const int ax = angle_to_sin_idx(v3_angle_x);

    const int sinY = sin_lookup(ay);
    const int cosY = cos_lookup(ay);
    const int sinX = sin_lookup(ax);
    const int cosX = cos_lookup(ax);

    /* Rotate about Y: x' = x*cos - z*sin; z' = x*sin + z*cos. */
    const int x1 = fixmul_sin(x0, cosY) - fixmul_sin(z0, sinY);
    const int z1 = fixmul_sin(x0, sinY) + fixmul_sin(z0, cosY);
    const int y1 = y0;

    /* Rotate about X: y'' = y*cos - z'*sin; z'' = y*sin + z'*cos. */
    const int y2 = fixmul_sin(y1, cosX) - fixmul_sin(z1, sinX);
    const int z2 = fixmul_sin(y1, sinX) + fixmul_sin(z1, cosX);
    const int x2 = x1;

    /* Perspective: screen = (coord * zoom) / (zoom + z + translation), then
     * shift to screen center. The web port noted that with vertices stored
     * at engine scale, zoom=1000 keeps object size in the visible window;
     * we use the same constant. */
    const int denom = v3_zoom + z2 + v3_translation;
    const int d = denom > 0 ? denom : 1;
    const int sx = ((x2 * v3_zoom) / d) + SCREEN_W / 2;
    const int sy = ((y2 * v3_zoom) / d) + SCREEN_H / 2;

    out->sx = (int16_t)sx;
    out->sy = (int16_t)sy;
    out->depth = (int16_t)(z2 + v3_translation);
}

void v3_normal_to_uv(const vec3_t *normal, int *u_out, int *v_out)
{
    /* Direct port of dawn_final.s:1693-1727 (the asm normal-rotation
     * loop). The TS reference (web_port/vector3d.ts:110-144) had one
     * bug we'd been inheriting:
     *
     *   It inverse-rotates normals via `Math.sin(-ay)`. The asm reuses
     *   the SAME d5/d6 sin/cos values as the vertex rotation at lines
     *   1632-1691 — normals rotate the same direction as the object.
     *   That gives world-anchored env-map lighting (highlight stays
     *   where the "light" is in world space, surface slides through);
     *   the TS gave object-anchored (highlight painted on the surface).
     *
     * Bias accounting: the asm stores `high16(rot_nx) + 64` to
     * tm_imag (line 1722) AND at sample time the `light` base pointer
     * is `ppic + 256*64 + 64` (line 1088), so the effective sample x
     * is `64 + u_stored + 64 = 128 + high16(rot_nx)`. We bake both
     * +64s into a single +128 here so env_sample() sees ready-to-look-
     * up coordinates. Brightness peaks at u=128, which happens when
     * `high16(rot_nx) = 0` — viewer-facing normals, the physically
     * correct location for a specular highlight.
     *
     * Our fixmul_sin uses >> 15 with sin/cos peaking at ~32255, which
     * preserves magnitude. The asm's `swap`/high16 pattern is >> 16,
     * giving half the magnitude — so `(nx1 >> 1)` here matches the
     * asm's `high16(rot_nx)` directly. The asm's `asl d2` doubling of
     * rotated_z before the X rotation is implicit in our path: our
     * nz1 already IS the asm's `2*high16(rot_z)`. */
    const int ay = angle_to_sin_idx(v3_angle_y);
    const int ax = angle_to_sin_idx(v3_angle_x);
    const int sinY = sin_lookup(ay);   /* POSITIVE — forward rotation */
    const int cosY = cos_lookup(ay);
    const int sinX = sin_lookup(ax);
    const int cosX = cos_lookup(ax);

    const int nx0 = normal->x;
    const int ny0 = normal->y;
    const int nz0 = normal->z;

    /* Forward Y rotation: x' = x cosY - z sinY,  z' = x sinY + z cosY. */
    const int nx1 = fixmul_sin(nx0, cosY) - fixmul_sin(nz0, sinY);
    const int nz1 = fixmul_sin(nx0, sinY) + fixmul_sin(nz0, cosY);

    /* Forward X rotation, only the y' we need for v. */
    const int ny2 = fixmul_sin(ny0, cosX) - fixmul_sin(nz1, sinX);

    *u_out = ((nx1 >> 1) + 128) & 0xFF;
    *v_out = ((ny2 >> 1) + 128) & 0xFF;
}

File: 02_Dawn/web_port/web/src/vector3d.ts
/**
 * 3D vector rotation and projection
 * Port of Amiga 3D transformation routines
 *
 * Original rotation: dawn_final.s:1632-1728
 * Original backface culling: dawn_final.s:1768-1822
 */

export interface Vector3D {
    x: number;
    y: number;
    z: number;
}

export interface Vector2D {
    x: number;
    y: number;
}

export interface ScreenCoord {
    x: number;  // Screen X (0-159)
    y: number;  // Screen Y (0-127)
    z: number;  // Depth for sorting
}

/**
 * 3D rotation and projection system
 */
export class Vector3DSystem {
    // Rotation angles (0-65535 range, divided by 32 for sine table)
    // Assembly stores 16-bit angles and uses "lsr #5" to get 0-1023 range
    angleX: number = 0;
    angleY: number = 0;
    angleZ: number = 0;

    // Translation distance (Z offset)
    // Assembly: RO_trans = 3000; we keep the same base distance.
    translation: number = 3000;

    // Perspective factor
    // Assembly: zoom = 1000 (constant at line 1630)
    // Must match original to maintain correct object scaling
    zoom: number = 1000;

    // Assembly uses lsr #5 to convert 16-bit angle to 0-1023 range
    private static readonly ANGLE_SCALE = (Math.PI * 2) / (1024 * 32);

    /**
     * Rotate and project a single point using floating point math.
     * Matches the 68k order: first rotate around Y, then around X.
     */
    rotateAndProject(point: Vector3D): ScreenCoord {
        const ay = this.angleY * Vector3DSystem.ANGLE_SCALE;
        const ax = this.angleX * Vector3DSystem.ANGLE_SCALE;

        const sinY = Math.sin(-ay);
        const cosY = Math.cos(ay);
        const sinX = Math.sin(-ax);
        const cosX = Math.cos(ax);

        // Rotate around Y (uses original X and Z)
        const x1 = point.x * cosY - point.z * sinY;
        const z1 = point.z * cosY + point.x * sinY;
        const y0 = point.y;

        // Rotate around X
        const y1 = y0 * cosX - z1 * sinX;
        const z2 = z1 * cosX + y0 * sinX;
        const x2 = x1;

        // Perspective projection
        // Assembly formula (dawn_final.s:1676-1686):
        //   screenX = (x >> 8) / (z + translation + zoom) + 80
        //   which is: (x / 256) / denominator + center
        //
        // TypeScript vertices are scaled by POSITION_SCALE = 1/8 (torus.ts:12)
        // The assembly shift right by 8 is a fixed-point conversion.
        // After rotation, assembly has 32-bit values that need >>8 before division.
        //
        // To match assembly projection with our floating point and 1/8 vertex scale:
        // We need to compensate for the vertex scale and match the division pattern.
        // The working formula uses: x * zoom / (zoom + z + translation)
        // With zoom=1000 (original value), this closely approximates the assembly behavior.
        const depth = z2 + this.translation;
        const denom = this.zoom + depth;
        const scale = this.zoom / (denom !== 0 ? denom : 1);

        const screenX = Math.round(x2 * scale + 80);
        const screenY = Math.round(y1 * scale + 64);

        return { x: screenX, y: screenY, z: depth };
    }

    /**
     * Rotate and project array of points
     */
    transformPoints(points: Vector3D[], output: ScreenCoord[]): void {
        for (let i = 0; i < points.length; i++) {
            output[i] = this.rotateAndProject(points[i]);
        }
    }

    /**
     * Rotate a normal vector (for environment mapping)
     * Original: dawn_final.s:1693-1727
     *
     * Rotates normals in INVERSE direction to keep lighting from viewer direction
     * as object rotates.
     */
    rotateNormalToUV(normal: Vector3D): { u: number; v: number } {
        const ax = this.angleX * Vector3DSystem.ANGLE_SCALE;
        const ay = this.angleY * Vector3DSystem.ANGLE_SCALE;

        // The matrix below already applies the inverse rotation, so we keep the original angles.
        const sinX = Math.sin(-ax);
        const cosX = Math.cos(ax);
        const sinY = Math.sin(-ay);
        const cosY = Math.cos(ay);

        // Apply inverse rotations
        let nx = normal.x;
        let ny = normal.y;
        let nz = normal.z;

        // Inverse Y rotation
        const nx1 = nx * cosY - nz * sinY;
        const nz1 = nx * sinY + nz * cosY;
        const ny1 = ny;

        // Inverse X rotation with Z doubling
        const nz1_doubled = nz1 * 1;
        const ny2 = (ny1 * cosX - nz1_doubled * sinX) / 1.0;
        const nx2 = nx1 / 1.0;

        // Add 64 offset (line 1721-1725) while preserving fractional precision
        // Fractional bits mimic the 68k fixed-point format.

        const u = Math.round(nx2 * 64 + 64) & 0xFF;
        const v = Math.round(ny2 * 64 + 64) & 0xFF;
        // const u = Math.round(nx2 ) & 0xFF;
        // const v = Math.round(ny2 ) & 0xFF;

        return { u, v };
    }

    /**
     * Calculate cross product for backface culling
     * Original: dawn_final.s:1782-1800
     *
     * Returns: positive if visible, negative/zero if backfacing
     */
    static crossProduct2D(
        p1: ScreenCoord,
        p2: ScreenCoord,
        p3: ScreenCoord
    ): number {
        // Create two edge vectors
        const v1x = p1.x - p2.x;
        const v1y = p1.y - p2.y;
        const v2x = p3.x - p2.x;
        const v2y = p3.y - p2.y;

        // Cross product (Z component)
        // v1x * v2y - v1y * v2x
        return v2x * v1y - v1x * v2y;
    }

    /**
     * Calculate 3D cross product for normal vectors
     * Original: dawn_final.s:819-846
     */
    static crossProduct3D(
        v1: Vector3D,
        v2: Vector3D
    ): Vector3D {
        // v1 x v2 = (v1y*v2z - v1z*v2y, v1z*v2x - v1x*v2z, v1x*v2y - v1y*v2x)
        const nx = ((v1.y * v2.z) - (v1.z * v2.y)) >> 8;
        const ny = ((v1.z * v2.x) - (v1.x * v2.z)) >> 8;
        const nz = ((v1.x * v2.y) - (v1.y * v2.x)) >> 8;

        return { x: nx, y: ny, z: nz };
    }
}
